### PR TITLE
[TOOLS-2405] KPIDataService Refactoring

### PR DESCRIPTION
**Pull request by heronoyama**
- Create KpiDataService;
- Change the flow for all Kpi's, using Snaphost or not;
- Change the flow for FollowupFacade to get Snapshot through KpiDataService;
- Remove a indirection existent on FollowupFacade;
- Rellocation and creation of new tests;
- Relocation of TouchTime Filter.

### CODE REVIEW

File: kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/enviroment/MockedServices.java [hunk 1]
```diff
@@ -301,7 +260,36 @@ private void mockService() {
...
+            return service;
+        }
+        
+        private void mock() {
+            service = Mockito.mock(KpiDataService.class);
+            Map<String, Map<KpiLevel, List<IssueKpi>>> byProject = issuesRepository.getIssuesByProject();
+            byProject.entrySet().stream().forEach(entry -> {
+                String projectKey = entry.getKey();
+                Map<KpiLevel, List<IssueKpi>> issuesByProject = entry.getValue();
+                issuesByProject.entrySet().forEach(levelIssues -> mockIssuesByLevel(projectKey,levelIssues.getKey(),levelIssues.getValue()));;
```
Comment (objective-solutions-user) [inline kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/enviroment/MockedServices.java]: MINOR Remove this empty statement. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AEmptyStatementUsageCheck)

File: kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/enviroment/MockedServices.java [hunk 2]
```diff
@@ -84,83 +97,53 @@ public MetadataServiceMocker metadata() {
...
             byProject.entrySet().stream().forEach(entry -> {
                 String projectKey = entry.getKey();
-                List<IssueKpi> issuesByProject = entry.getValue().stream()
-                                                        .map(IssueKpiMocker::buildIssueKpi)
-                                                        .collect(Collectors.toList());
-                saveIssues(issuesByProject);
-                Mockito.when(service.getIssuesFromCurrentState(projectKey, environment.getTimezone(), level))
-                    .thenReturn(issuesByProject);
+                Map<KpiLevel, List<IssueKpi>> issuesByProject = entry.getValue();
+                issuesByProject.entrySet().forEach(levelIssues -> mockIssuesByLevel(projectKey,levelIssues.getKey(),levelIssues.getValue()));;
```
Comment (objective-solutions-user) [inline kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/enviroment/MockedServices.java]: MINOR Remove this empty statement. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AEmptyStatementUsageCheck)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/KpiDataServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        private String projectKey;
+
+        public FollowupSnapshotBehavior(String projectKey) {
+            this.projectKey = projectKey;
+        }
+
+        @Override
+        public void behave(KpiEnvironment environment) {
+
+            GenerateSnapshot datasetFactory = new GenerateSnapshot(environment);
```
Comment (jhonypg) [inline kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/KpiDataServiceTest.java]: Talvez mudar o nome da var para algo mais próximo da classe.

File: kpi-parent/kpi-wip/src/test/java/objective/taskboard/followup/WipKPIServiceTest.java [hunk 4]
```diff
@@ -124,9 +124,9 @@ private DSLKpi dsl() {
 
         @Override
         public void behave(KpiEnvironment environment) {
-            GenerateAnalyticsDataSets datasetFactory = new GenerateAnalyticsDataSets(environment);
+            GenerateSnapshot datasetFactory = new GenerateSnapshot(environment);
```
Comment (jhonypg) [inline kpi-parent/kpi-wip/src/test/java/objective/taskboard/followup/WipKPIServiceTest.java]: Talvez mudar o nome da var para algo mais próximo da classe.

File: kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/enviroment/snapshot/GenerateSnapshot.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import objective.taskboard.followup.FollowUpSnapshot;
+import objective.taskboard.followup.FromJiraDataSet;
+import objective.taskboard.followup.KpiHelper;
+import objective.taskboard.followup.SyntheticTransitionsDataSet;
+import objective.taskboard.followup.kpi.enviroment.KpiEnvironment;
+
+public class GenerateSnapshot {
+    
+    private Optional<FromJiraDataSet> fromJiraDs = Optional.empty();
+    private GenerateAnalyticsDataSets analyticBuilder;
```
Comment (jhonypg) [inline kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/enviroment/snapshot/GenerateSnapshot.java]: O nome "builder" para o GenerateAnalyticsDataSets acaba ficando um pouco estranho, já que ele só tem alguns gets.

File: kpis/src/main/java/objective/taskboard/followup/kpi/KpiDataService.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    public KpiDataService(IssueKpiService issueKpiService, FollowUpSnapshotService followupSnapshotService) {
+        this.issueKpiService = issueKpiService;
+        this.followupSnapshotService = followupSnapshotService;
+    }
+    
+    public List<IssueKpi> getIssuesFromCurrentState(String projectKey, ZoneId timezone, KpiLevel kpiLevel){
+        return issueKpiService.getIssuesFromCurrentState(projectKey, timezone, kpiLevel);
+    }
+
+    public FollowUpSnapshot getSnapshotFromCurrentState(ZoneId timezone,String projectKey) {
```
Comment (jhonypg) [inline kpis/src/main/java/objective/taskboard/followup/kpi/KpiDataService.java]: Como o método acima está com os parâmetros na sequencia 1  projectKey, 2 timezone e 3 kpiLevel, poderia deixar esse método na mesma sequência (1 e 2)

File: kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/KpiDataServiceTest.java [hunk 7]
```diff
@@ -160,20 +354,48 @@ public ServeIssuesFromCurrentState(String projectKey, KpiLevel level) {
...
+        @Override
+        protected List<IssueKpi> getIssues(KpiDataService subject, ZoneId timezone) {
+            return subject.getIssuesFromCurrentState(projectKey, timezone, level);
+        }
+        
+    }
+    
+    private class ServeIssuesFilteringByProject extends ServeIssuesFromCurrentState {
+    
+    private ServeIssuesFilteringByProject(String projectKey, KpiLevel level) {
```
Comment (jhonypg) [inline kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/KpiDataServiceTest.java]: Tá faltando espaço tab.

File: kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/KpiDataServiceTest.java [hunk 8]
```diff
@@ -160,20 +354,48 @@ public ServeIssuesFromCurrentState(String projectKey, KpiLevel level) {
...
+            super(projectKey, level);
+        }
+
+    @Override
+    protected List<IssueKpi> getIssues(KpiDataService subject, ZoneId timezone) {
```
Comment (jhonypg) [inline kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/KpiDataServiceTest.java]: Tá faltando espaço tab.

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [IssueKpiMocker.java#L271](https://github.com/objective-solutions/taskboard/blob/a7ef46fa2dc3abd90432e44d23fa27feaf3edbce/kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/services/IssueKpiMocker.java#L271): Remove this null-check of an "Optional". [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2789)
1. ![MAJOR][MAJOR] [IssueKpiMocker.java#L274](https://github.com/objective-solutions/taskboard/blob/a7ef46fa2dc3abd90432e44d23fa27feaf3edbce/kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/services/IssueKpiMocker.java#L274): A "NullPointerException" could be thrown; "type" is nullable here. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2259)
1. ![MAJOR][MAJOR] [TransitionsBuilder.java#L49](https://github.com/objective-solutions/taskboard/blob/a7ef46fa2dc3abd90432e44d23fa27feaf3edbce/kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/services/TransitionsBuilder.java#L49): Call "Optional#isPresent()" before accessing the value. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS3655)
1. ![MAJOR][MAJOR] [TransitionsBuilder.java#L54](https://github.com/objective-solutions/taskboard/blob/a7ef46fa2dc3abd90432e44d23fa27feaf3edbce/kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/services/TransitionsBuilder.java#L54): A "NullPointerException" could be thrown; "transition" is nullable here. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2259)
1. ![MAJOR][MAJOR] [TransitionsBuilder.java#L98](https://github.com/objective-solutions/taskboard/blob/a7ef46fa2dc3abd90432e44d23fa27feaf3edbce/kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/services/TransitionsBuilder.java#L98): Call "Optional#isPresent()" before accessing the value. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS3655)
1. ![MINOR][MINOR] [ExceptionBehavior.java#L19](https://github.com/objective-solutions/taskboard/blob/a7ef46fa2dc3abd90432e44d23fa27feaf3edbce/kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/services/ExceptionBehavior.java#L19): Remove this empty statement. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AEmptyStatementUsageCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**